### ISSUE TITLE
[Feat] MapboxLayer / MapboxOverlay integration with `globe` projection

### ISSUE THREAD

**Issue by Beilinson** [visgl/deck.gl#7920]
### Target Use Case

In my organization we are required to use the Globe projection for our maps. Recently I stumbled upon the deck.gl library and would love to use it for the potential in additional 3D visualization it provides on top of mapbox.

However we quickly realized that although mapbox supports custom layers in globe view, the deck.gl layers don't update according to the projection which provides for some silly looking bugs like this one:

https://github.com/visgl/deck.gl/assets/54665621/b82f7be5-6273-43b2-ad0d-2cf997970243

We'd love to contribute to the deck.gl library and get this behavior fixed, where should we look to begin with?

### Proposal

The mapbox `globe` projection is an [adaptive projection](https://docs.mapbox.com/mapbox-gl-js/guides/projections/#adaptive-projection-behavior) which gradually morphs from the 3d globe to a 2d mercator projection between zoom levels 6-8. Seeing as these are arbitrary numbers which could be changed in the future, the MapboxLayer / MapboxOverlay integration should always update its projection matrix (and any other dependent properties) by the current zoom projection so that it stays accurate. All points without height should also clamp to the spheroid, while datasources with height should be offset from the spheroid rather than the 2d plane in mercator projection.

**Comment by woozyking** [comment]
This would be amazing to have!

**Comment by chrisgervang** [comment]
We have some related issues, like https://github.com/visgl/react-map-gl/issues/2091. We need mapbox to expose more of their camera parameters to pull this off. See more discussion in https://github.com/mapbox/mapbox-gl-js/issues/12040. In particular:
> The only way I can make Mapbox integrations forward-compatible, by eliminating all hacks, is if `transform` can be consistently reproduced via public APIs.

Also, @Pessimistress accomplished this in `maplibre` with https://github.com/maplibre/maplibre-gl-js/pull/2535.. it'd be excellent if `mapbox` could implement this as well. It might not completely cover projections, but this is probably a pre-requisite...

**Comment by TtangBro** [comment]
> 
Hello, can I learn the implementation code of the effect of your comment video

**Comment by kubbo** [comment]
Is the feature "MapboxLayer / MapboxOverlay integration with globe projection" planned for feture iterations  ? 
When can we expect it to be implemented?

**Comment by Pessimistress** [comment]
Mapbox does not currently support custom layers in globe projection. We can't do anything until they expose sufficient API for us to implement integration.

**Comment by tangzijun** [comment]
> Mapbox does not currently support custom layers in globe projection. We can't do anything until they expose sufficient API for us to implement integration.

maplibre-gl@5.0.0-pre.3 has released globe projection and custom layer related APIs. Do we have any plans?

**Comment by Pessimistress** [comment]
For maplibre integration please refer to https://github.com/visgl/deck.gl/issues/9199

At the moment, we do not have any plan to support mapbox-gl's non-mercator projection. deck.gl does not perform projection on the CPU like their custom layer examples do. In order to be compatible, we will need to generalize deck.gl's own projection system to cover mapbox-gl configurations, but Mapbox's license explicitly prohibit us from learning from their source code.

**Comment by Pessimistress** [comment]
This is shipped in v9.1.

**Comment by yuriy-sng** [comment]
Hi @Pessimistress just to confirm: does deck.gl support Mapbox globe projection? Both in the What's New page of v9.1 and GH release all I see is MapLibre mentions. I could open an issue, but first trying to confirm I understood this right.

**Comment by LukasJK** [comment]
No, does not seem to work with mapbox

**Comment by chrisgervang** [comment]
Correct. Mapbox is closed sourced - bound by license, we need to treat it like a black box and only use public APIs. 

They haven't added the APIs deck would need to match their projection. We've asked many times and @Pessimistress even supplied them with a reference implantation via MapLibre. We'd love to support this one day - the ball is in Mapbox's court.

**Comment by MikhailSorokin** [comment]
It is quite frustrating that Mapbox does not have support for this. In my project, my deck.gl layers maps properly in a globe when in the mercator projection but it is wrong obviously since it doesn't rotate like a globe does. Once we switch to try this in Mapbox's globe projection, I know deck.gl can't do this, as mentioned in this thread, and it causes flicker in the layers. 

I think we need to put more pressure on mapbox to try to support this because this projection just becomes a straight up bug there then.
